nexus/src/app/update.rs

```diff
@@ -149,4 +157,170 @@ impl super::Nexus {
+        // it can still have an Unspecified release_source
+        let db_target_release =
+            self.datastore().target_release_get_current(opctx).await?;
+        let target_release = match db_target_release.tuf_repo_id {
+            Some(tuf_repo_id) => {
+                let version = self
+                    .datastore()
+                    .tuf_repo_get_version(opctx, &tuf_repo_id)
+                    .await?;
+                Some(views::TargetRelease {
+                    time_requested: db_target_release.time_requested,
+                    version,
+                })
+            }
+            None => None,
+        };
+
+        let components_by_release_version =
+            self.component_version_counts(opctx, &db_target_release).await?;
+
+        let time_last_progress = self
+            .datastore()
+            .blueprint_target_get_current(opctx)
+            .await?
+            .time_made_target;
+
+        Ok(views::UpdateStatus {
+            target_release: Nullable(target_release),
+            components_by_release_version,
+            time_last_progress,
+        })
+    }
+
+    /// Get component status using read-only queries to avoid batch operations
+    async fn component_version_counts(
+        &self,
+        opctx: &OpContext,
+        target_release: &nexus_db_model::TargetRelease,
+    ) -> Result<BTreeMap<String, usize>, Error> {
+        // Get the latest inventory collection
+        let Some(inventory) =
+            self.datastore().inventory_get_latest_collection(opctx).await?
+        else {
+            // No inventory collection available, return empty counts
+            return Ok(BTreeMap::new());
+        };
+
+        // Build current TargetReleaseDescription, defaulting to Initial if
+        // there is no tuf repo ID which, based on DB constraints, happens if
+        // and only if target_release_source is 'unspecified', which should only
+        // happen in the initial state before any target release has been set
+        let curr_target_desc = match target_release.tuf_repo_id {
+            Some(id) => TargetReleaseDescription::TufRepo(
+                self.datastore()
+                    .tuf_repo_get_by_id(opctx, id.into())
+                    .await?
+                    .into_external(),
+            ),
+            None => TargetReleaseDescription::Initial,
+        };
+
+        // Get previous target release (if it exists). Build the "prev"
+        // TargetReleaseDescription from the previous generation if available,
+        // otherwise fall back to Initial.
+        let prev_repo_id =
+            if let Some(prev_gen) = target_release.generation.prev() {
+                self.datastore()
+                    .target_release_get_generation(opctx, Generation(prev_gen))
+                    .await
+                    .internal_context("fetching previous target release")?
+                    .and_then(|r| r.tuf_repo_id)
```

Comment:
This feels a little sketchy?  I'm not sure.

Example: after a MUPdate, you might set the target release to the same value it was before.  Then the previous row will not be the previous release.

Today, the PlanningInput has this information.  Maybe `PlanningInputFromDb` has a better way to do this?

Reply:
Unfortunately, that is where I copied this logic from.

https://github.com/oxidecomputer/omicron/blob/4d5bdc6d348b27761348d763c4085f060bcefc18/nexus/reconfigurator/preparation/src/lib.rs#L163-L199

I certainly see your point, though, and it is a problem. In the internal API logic I'm borrowing, it seems like we will only turn up component versions that match either the old or the new one passed in, because we're matching against artifact hashes. In the MUPdate case where current and prev release are the same, components that are on the the MUPdate's version will all show up as `TufRepoVersion::Unknown`.

https://github.com/oxidecomputer/omicron/blob/4d5bdc6d348b27761348d763c4085f060bcefc18/nexus/types/src/internal_api/views.rs?plain=1#L937-L968

Reply:
Yeah, I think the upshot of the discussion in chat was that since the internal API logic doesn't do any better, it makes sense to do it this way for now and improve both later.